### PR TITLE
[NPU] Add NPU support for unit test

### PR DESCRIPTION
**Pull request by RUAN-ZX**
Unit tests would fail or skip when device=npu, and we definitely want to test all these wonderful features by official unit tests.
Here comes the commit to add NPU support for unit test. P.S. see what we have already done #4567.

**What I do in this commit**
1. Just add npu logic branch 
feat: Add npu support for skip_on_arch in tests/unit/util.py
feat: Add npu support for skip_on_cuda in tests/unit/util.py
feat: Add npu support for tests/unit/common.py

2. Set_device of accelerator before deepspeed.init_distributed in tests/unit/common.py
It would be friendlier and easier for other device like npu, if we can set_device of accelerator before init_distributed. Plus, setting device param before init sounds more reasonable.

3. Solve the problem of calling get_accelerator().random().fork_rng with non-cuda device
Function `train_cifar()` in `tests/unit/alexnet_model.py` calls `get_accelerator().random().fork_rng` without passing `device_type` explicitly. Unfortunately, `torch.random.fork_rng()` has default value setting `device_type=cuda` and non-cuda devices would fail to run. So my solution is explicitly passing `device_type=get_accelerator().device_name()`, and either cuda or non-cuda devices would perform correctly.

### CODE REVIEW

**Comment by RUAN-ZX** [comment]
@microsoft-github-policy-service agree

**Comment by hipudding** [comment]
Please add cuda check in "check_environment" to avoid warning meesage when using NPU as the backend.
Please squash these commits into one(use git rebase -i), and describe all the changes.

**Comment by RUAN-ZX** [comment]
@tjruwase Would you be so kind to review this commit, since we have some other commits based on this? Or maybe you can invite other reviewers to do the job? Thank you.

**Review by tjruwase — APPROVED** [review]

**Comment by RUAN-ZX** [comment]
@tjruwase It seems that the CI problem might be fixed by loadams in #4590? If so, would you please launch CI for me again?
 Several PRs (#4588, #4585, #4578, etc) have met the same problem as below:
`unit/inference/test_inference.py::TestMPSize::test[fp16-bloom] FAILED    [ 91%]`

**Comment by RUAN-ZX** [comment]
@tjruwase Perhaps #4591 has solved the problem of CI? Cause I see the latest commit #4598 manages to pass. Would you launch CI for me again?

**Comment by RUAN-ZX** [comment]
> @tjruwase Perhaps #4591 has solved the problem of CI? Cause I see the latest commit #4598 manages to pass. Would you launch CI for me again?

@tjruwase Could you launch CI for me again? The last two commits have already pass.

**Comment by RUAN-ZX** [comment]
@tjruwase Would you please launch CI for this PR? We have been waiting for a long time. Thanks :)

**Comment by tjruwase** [comment]
@RUAN-ZX, apologies for the delay in merging this PR. We had to push out a scheduled release last week. Unfortunately, I notice a new CI failure https://github.com/microsoft/DeepSpeed/actions/runs/6785863907/job/18445092642?pr=4569. 

Can you please take a look? It seems to have been caused by a recent merge. I will play close attention to ensure this PR is merged this week. Thanks for your patience.

**Comment by RUAN-ZX** [comment]
> @RUAN-ZX, apologies for the delay in merging this PR. We had to push out a scheduled release last week. Unfortunately, I notice a new CI failure https://github.com/microsoft/DeepSpeed/actions/runs/6785863907/job/18445092642?pr=4569.
> 
> Can you please take a look? It seems to have been caused by a recent merge. I will play close attention to ensure this PR is merged this week. Thanks for your patience.

Thank you! I found that an assert error: `CUDA_HOME does not exist, unable to compile CUDA op(s)` from op_builder/builder.py, and eventually it uses `cuda_home = os.environ.get('CUDA_HOME') or os.environ.get('CUDA_PATH')` to get `CUDA_HOME` in `torch\utils\cpp_extension.py`.
There is no code that may change this env param and I don't understand why `CUDA_HOME` is None :)

File: tests/unit/util.py [hunk 1]
```diff
@@ -26,6 +28,8 @@ def skip_on_cuda(valid_cuda):
         CUDA_VERSION = (CUDA_MAJOR * 10) + CUDA_MINOR
         if valid_cuda.count(CUDA_VERSION) == 0:
             pytest.skip(f"requires cuda versions {valid_cuda}")
+    elif deepspeed.accelerator.get_accelerator().device_name() == 'npu':
+        return
     else:
         assert deepspeed.accelerator.get_accelerator().device_name() == 'xpu'
```
Comment (loadams) [inline tests/unit/util.py:34]: What is the difference between an xpu and an npu here?

Comment (RUAN-ZX) [reply tests/unit/util.py]: Well at first we want to add npu version control logic here. But since we can't show you the code right now, it seems to be confusing and over-engineering, 
I think I have better to merge npu and xpu branches by assert device_name in ['npu','xpu']. You can see changes in latest [commit](https://github.com/microsoft/DeepSpeed/pull/4569/commits/4e0b9aa97aa217335beb9d9b2af66233f6f88b69).

Thank you for your review :)

Comment (hipudding) [reply tests/unit/util.py]: They are different backend devices. Ascend NPU is a series of AI processors.
please refer to [hiascend](https://www.hiascend.com/) for details.

File: tests/unit/util.py [hunk 2]
```diff
@@ -15,7 +15,7 @@ def skip_on_arch(min_arch=7):
         if torch.cuda.get_device_capability()[0] < min_arch:  #ignore-cuda
             pytest.skip(f"needs higher compute capability than {min_arch}")
     else:
-        assert deepspeed.accelerator.get_accelerator().device_name() == 'xpu'
+        assert deepspeed.accelerator.get_accelerator().device_name() in ['xpu', 'npu']
```
Comment (tjruwase) [inline tests/unit/util.py:18]: Can we not use [DS_ACCELERATOR_LIST](https://github.com/microsoft/DeepSpeed/blob/1d1a20c5a1dd980438393500f14d8dc188f6258a/accelerator/real_accelerator.py#L53C10-L53C10) here?

Comment (RUAN-ZX) [reply tests/unit/util.py:18]: You are right, I will make [DS_ACCELERATOR_LIST](https://github.com/microsoft/DeepSpeed/blob/1d1a20c5a1dd980438393500f14d8dc188f6258a/accelerator/real_accelerator.py#L53C10-L53C10) as a public-accessible valid accelerator list, and we can check if current device is valid or not in all non-cuda logic branches.

Comment (RUAN-ZX) [reply tests/unit/util.py]: @tjruwase I expose `is_current_accelerator_supported()` in [commit](https://github.com/microsoft/DeepSpeed/pull/4569/commits/ee399db7655f6fc441edbcd9a91e8d9652970fc2) and I use it to assert.

**Comment by RUAN-ZX** [comment]
@tjruwase I have solved problems raised by pre-commit hooks, please launch CI again :)

File: accelerator/real_accelerator.py [hunk 3]
```diff
@@ -34,14 +36,18 @@ def _validate_accelerator(accel_obj):
     # accelerator.abstractor_accelerator
     # or deepspeed.accelerator.abstract_accelerator, consider accel_obj
     # is a conforming object
-    if not ((dsa1 != None and isinstance(accel_obj, dsa1)) or (dsa2 != None and isinstance(accel_obj, dsa2))):
+    if not ((dsa1 is not None and isinstance(accel_obj, dsa1)) or (dsa2 is not None and isinstance(accel_obj, dsa2))):
```
Comment (tjruwase) [inline accelerator/real_accelerator.py:39]: Why is this change needed? How is it different from previous code?

Comment (tjruwase) [reply accelerator/real_accelerator.py:39]: So, I found justification [here](https://peps.python.org/pep-0008/#programming-recommendations). Thanks for this code improvement!.

**Review by tjruwase — APPROVED** [review]

File: accelerator/real_accelerator.py (hunk 3 above)
Comment (RUAN-ZX) [reply accelerator/real_accelerator.py:39]: Happy to!.
